File: 20-actors/matsurigoto/methods/modules/civil_registry.py
#!/usr/bin/env python3
"""matsurigoto 政 — `civil-registry` module (R0 reference implementation).

ADR-2606052300. Second executable vertical slice: the CRVS (Civil Registration & Vital
Statistics) engine behind `civil.birth.register` / `civil.death.register` /
`civil.marriage.register` / `residency.move-in` / `residency.certificate` — i.e. 住所管理・戸籍.

WHAT IT IS: pure-function VALIDATION + APPEND-ONLY RECORD CONSTRUCTION for vital events,
shaped on the UN CRVS principles + OpenCRVS data model. A registration is validated (a birth
needs a child + ≥1 parent + a non-future occurrence; a marriage needs two DISTINCT, unmarried
partners; a death needs a decedent), then emitted as an immutable kotoba datom + an UNSIGNED
certificate skeleton (a W3C VC the governing organ signs with ITS own key).

Charter invariants (nusa/tazuna/ake pattern), why 住所管理 is charter-clean here:
  G1 no-operator-master-key : SERVER_HELD_AUTHORITY False; certificates returned UNSIGNED — the
                              Council (sovereign) / adopting state signs, never this module.
  G2 spec-derived-only      : UN CRVS + OpenCRVS + W3C VC 2.0 shapes only.
  G5 append-only (非終末論)  : every helper RETURNS A NEW record list; nothing is overwritten.
                              A correction is itself an appended record (kotoba-canonical
                              ADR-2605312345 + ake G5). No :delete / :overwrite exists.
  G6 data-minimization      : only the fields the vital event requires; no surveillance surface.

stdlib only, no I/O, no network. Importable as a kotoba-wasm module contract.
"""
from __future__ import annotations

# G1: this module holds NO signing authority and signs no certificate.
SERVER_HELD_AUTHORITY = False

_VITAL_KINDS = {"birth", "death", "marriage"}


def _iso(s: str) -> str:
    """ISO-8601 strings sort lexically; we only need ordering + non-future checks (no parsing lib)."""
    if not isinstance(s, str) or len(s) < 4 or not s[:4].isdigit():
        raise ValueError(f"timestamp must be ISO-8601, got {s!r}")
    return s


def _unsigned_certificate(kind: str, subject: str, record_id: str) -> dict:
    """A W3C-VC certificate SKELETON. G1: unsigned — the governing organ signs with ITS key."""
    return {
        "@context": ["https://www.w3.org/ns/credentials/v2"],
        "type": ["VerifiableCredential", f"{kind.capitalize()}Certificate"],
        "credentialSubject": {"id": subject, "record": record_id},
        "proof": None,                                   # G1 — this module signs nothing
        "server_held_authority": SERVER_HELD_AUTHORITY,  # False
        "status": "issued-unsigned",
    }


def _record(kind: str, record_id: str, fields: dict, occurred_at: str) -> dict:
    """An immutable CRVS datom (append-only). 非終末論: this is one event, never a final state."""
    return {
        "record_id": record_id,
        "vital_kind": kind,
        "occurred_at": occurred_at,
        "fields": dict(fields),       # data-minimized (G6)
        "immutable": True,            # G5 — appended, never overwritten
    }


def register_birth(record_id: str, child: str, parents: list, place: str,
                   occurred_at: str, now: str) -> dict:
    """Validate + construct a birth registration (UN CRVS). Pure function.

    Requires a child, at least one parent, a place, and a non-future occurrence.
    Returns the immutable record + an unsigned birth certificate.
    """
    if not child:
        raise ValueError("birth: child is required")
    if not parents:
        raise ValueError("birth: at least one parent is required")
    if not place:
        raise ValueError("birth: place is required")
    if _iso(occurred_at) > _iso(now):
        raise ValueError("birth: occurrence cannot be in the future")
    rec = _record("birth", record_id,
                  {"child": child, "parents": list(parents), "place": place}, occurred_at)
    return {"record": rec, "certificate": _unsigned_certificate("birth", child, record_id)}


def register_death(record_id: str, decedent: str, place: str,
                   occurred_at: str, now: str, cause: str | None = None) -> dict:
    """Validate + construct a death registration (UN CRVS). Pure function."""
    if not decedent:
        raise ValueError("death: decedent is required")
    if not place:
        raise ValueError("death: place is required")
    if _iso(occurred_at) > _iso(now):
        raise ValueError("death: occurrence cannot be in the future")
    fields = {"decedent": decedent, "place": place}
    if cause:
        fields["cause"] = cause  # ICD-11 coded where present
    rec = _record("death", record_id, fields, occurred_at)
    return {"record": rec, "certificate": _unsigned_certificate("death", decedent, record_id)}


def register_marriage(record_id: str, partner_a: str, partner_b: str, place: str,
                      occurred_at: str, now: str, existing_marriages: tuple = ()) -> dict:
    """Validate + construct a marriage registration (UN CRVS). Pure function.

    Requires two DISTINCT partners, a place, a non-future occurrence, and that neither partner
    is already in an active marriage within `existing_marriages` (a tuple of frozensets/pairs).
    """
    if not partner_a or not partner_b:
        raise ValueError("marriage: two partners are required")
    if partner_a == partner_b:
        raise ValueError("marriage: partners must be distinct")
    if not place:
        raise ValueError("marriage: place is required")
    if _iso(occurred_at) > _iso(now):
        raise ValueError("marriage: occurrence cannot be in the future")
    already = {p for pair in existing_marriages for p in pair}
    if partner_a in already or partner_b in already:
        raise ValueError("marriage: a partner is already in an active marriage")
    rec = _record("marriage", record_id,
                  {"partners": sorted([partner_a, partner_b]), "place": place}, occurred_at)
    return {"record": rec, "certificate": _unsigned_certificate("marriage", record_id, record_id)}


def register_residency(record_id: str, person: str, new_address: str,
                       occurred_at: str, now: str, prior_address: str | None = None) -> dict:
    """Residence registration (転入届). Append-only — a move-in is a new datom, the prior
    address is retained in history (非終末論), never overwritten (G5)."""
    if not person:
        raise ValueError("residency: person is required")
    if not new_address:
        raise ValueError("residency: new_address is required")
    if _iso(occurred_at) > _iso(now):
        raise ValueError("residency: occurrence cannot be in the future")
    fields = {"person": person, "address": new_address}
    if prior_address:
        fields["prior_address"] = prior_address
    rec = _record("residency", record_id, fields, occurred_at)
    return {"record": rec, "certificate": _unsigned_certificate("residency", person, record_id)}


def append(history: list, result: dict) -> list:
    """G5: append a registration to a history, returning a NEW list (never mutate in place)."""
    return list(history) + [result["record"]]


def current_address(history: list, person: str) -> str | None:
    """Latest residency datom for a person = current address (max occurred_at). 非終末論."""
    fixes = [r for r in history if r["vital_kind"] == "residency" and r["fields"]["person"] == person]
    if not fixes:
        return None
    return max(fixes, key=lambda r: r["occurred_at"])["fields"]["address"]


def solve(*_args, **_kwargs):
    raise RuntimeError(
        "civil-registry R0: reference validation + record construction only. Live registration "
        "against a real civil register is Council+operator gated (principal A: Council Lv7+; "
        "principal B: adopting state)."
    )


if __name__ == "__main__":
    b = register_birth("birth-1", "child:aoi", ["parent:rin"], "東京都新宿区",
                       "2026-06-01T09:00:00Z", "2026-06-05T00:00:00Z")
    print("birth record:", b["record"]["record_id"], "cert signed?", b["certificate"]["proof"])

File: 20-actors/matsurigoto/methods/modules/civil_registry.clj
;; civil_registry.clj — matsurigoto 政 `civil-registry` module (R0 reference implementation).
;;
;; Clojure port of civil_registry.py (ADR-2606062300), Wave 1 of the clj-native migration
;; (ADR-2606142300). The CRVS (Civil Registration & Vital Statistics) engine behind
;; civil.birth/death/marriage.register / residency.move-in / residency.certificate — 住所管理・戸籍.
;; Pure-function VALIDATION + APPEND-ONLY record construction shaped on UN CRVS + OpenCRVS:
;; a registration is validated (birth needs a child + ≥1 parent + non-future occurrence; a
;; marriage needs two DISTINCT, unmarried partners; a death needs a decedent), then emitted as
;; an immutable record + an UNSIGNED W3C-VC certificate skeleton (the governing organ signs).
;;
;;   G1 no-operator-master-key : server-held-authority false; certificates UNSIGNED.
;;   G2 spec-derived-only      : UN CRVS + OpenCRVS + W3C VC 2.0 shapes only.
;;   G5 append-only (非終末論)  : every helper RETURNS A NEW record list; nothing overwritten —
;;                               a correction is itself an appended record (ADR-2605312345).
;;   G6 data-minimization      : only the fields the vital event requires.
;;
;; stdlib only, no I/O, no network. ISO-8601 strings sort lexically (ordering + non-future only).
(ns matsurigoto.methods.modules.civil-registry
  (:require [clojure.string :as str]))

(def server-held-authority false)  ; G1

(defn- iso
  "Validate an ISO-8601 timestamp (year-leading); strings sort lexically for ordering/non-future."
  [s]
  (if (and (string? s) (>= (count s) 4) (every? #(Character/isDigit %) (subs s 0 4)))
    s
    (throw (ex-info (str "timestamp must be ISO-8601, got " (pr-str s)) {:ts s}))))

(defn- future? [occurred-at now]
  (pos? (compare (iso occurred-at) (iso now))))

(defn- unsigned-certificate
  [kind subject record-id]
  {:context               ["https://www.w3.org/ns/credentials/v2"]    ; JSON-LD "@context"
   :type                  ["VerifiableCredential" (str (str/capitalize kind) "Certificate")]
   :credential-subject    {:id subject :record record-id}
   :proof                 nil                                          ; G1
   :server-held-authority server-held-authority                       ; false
   :status                "issued-unsigned"})

(defn- record-of
  "An immutable CRVS record (append-only). 非終末論: one event, never a final state."
  [kind record-id fields occurred-at]
  {:record-id   record-id
   :vital-kind  kind
   :occurred-at occurred-at
   :fields      fields           ; data-minimized (G6)
   :immutable   true})           ; G5

(defn register-birth
  "Validate + construct a birth registration (UN CRVS). Requires child, ≥1 parent, place,
   non-future occurrence."
  [record-id child parents place occurred-at now]
  (when (str/blank? (str child)) (throw (ex-info "birth: child is required" {})))
  (when (empty? parents) (throw (ex-info "birth: at least one parent is required" {})))
  (when (str/blank? (str place)) (throw (ex-info "birth: place is required" {})))
  (when (future? occurred-at now) (throw (ex-info "birth: occurrence cannot be in the future" {})))
  {:record      (record-of "birth" record-id {:child child :parents (vec parents) :place place} occurred-at)
   :certificate (unsigned-certificate "birth" child record-id)})

(defn register-death
  "Validate + construct a death registration (UN CRVS). Optional ICD-11 `cause`."
  ([record-id decedent place occurred-at now] (register-death record-id decedent place occurred-at now nil))
  ([record-id decedent place occurred-at now cause]
   (when (str/blank? (str decedent)) (throw (ex-info "death: decedent is required" {})))
   (when (str/blank? (str place)) (throw (ex-info "death: place is required" {})))
   (when (future? occurred-at now) (throw (ex-info "death: occurrence cannot be in the future" {})))
   (let [fields (cond-> {:decedent decedent :place place} (and cause (seq (str cause))) (assoc :cause cause))]
     {:record      (record-of "death" record-id fields occurred-at)
      :certificate (unsigned-certificate "death" decedent record-id)})))

(defn register-marriage
  "Validate + construct a marriage registration (UN CRVS). Requires two DISTINCT partners, a
   place, a non-future occurrence, and that neither partner is already in an active marriage
   within `existing-marriages` (a coll of partner pairs)."
  ([record-id partner-a partner-b place occurred-at now]
   (register-marriage record-id partner-a partner-b place occurred-at now []))
  ([record-id partner-a partner-b place occurred-at now existing-marriages]
   (when (or (str/blank? (str partner-a)) (str/blank? (str partner-b)))
     (throw (ex-info "marriage: two partners are required" {})))
   (when (= partner-a partner-b) (throw (ex-info "marriage: partners must be distinct" {})))
   (when (str/blank? (str place)) (throw (ex-info "marriage: place is required" {})))
   (when (future? occurred-at now) (throw (ex-info "marriage: occurrence cannot be in the future" {})))
   (let [already (set (mapcat seq existing-marriages))]
     (when (or (already partner-a) (already partner-b))
       (throw (ex-info "marriage: a partner is already in an active marriage" {}))))
   {:record      (record-of "marriage" record-id
                            {:partners (vec (sort [partner-a partner-b])) :place place} occurred-at)
    :certificate (unsigned-certificate "marriage" record-id record-id)}))

(defn register-residency
  "Residence registration (転入届). Append-only — a move-in is a new record, the prior address
   is retained in history (非終末論), never overwritten (G5)."
  ([record-id person new-address occurred-at now] (register-residency record-id person new-address occurred-at now nil))
  ([record-id person new-address occurred-at now prior-address]
   (when (str/blank? (str person)) (throw (ex-info "residency: person is required" {})))
   (when (str/blank? (str new-address)) (throw (ex-info "residency: new_address is required" {})))
   (when (future? occurred-at now) (throw (ex-info "residency: occurrence cannot be in the future" {})))
   (let [fields (cond-> {:person person :address new-address}
                  (and prior-address (seq (str prior-address))) (assoc :prior-address prior-address))]
     {:record      (record-of "residency" record-id fields occurred-at)
      :certificate (unsigned-certificate "residency" person record-id)})))

(defn append
  "G5: append a registration to a history, returning a NEW vector (never mutate in place)."
  [history result]
  (conj (vec history) (:record result)))

(defn current-address
  "Latest residency record for a person = current address (max occurred-at). 非終末論."
  [history person]
  (let [moves (filter #(and (= "residency" (:vital-kind %)) (= person (get-in % [:fields :person]))) history)]
    (when (seq moves)
      ;; occurred-at is an ISO string → compare lexically (max-key needs Numbers); first-on-tie like Python max.
      (get-in (reduce (fn [a b] (if (pos? (compare (:occurred-at b) (:occurred-at a))) b a)) moves)
              [:fields :address]))))

(defn solve
  [& _]
  (throw (ex-info (str "civil-registry R0: reference validation + record construction only. Live "
                       "registration against a real civil register is Council+operator gated "
                       "(principal A: Council Lv7+; principal B: adopting state).")
                  {:gated true})))
